src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java

```diff
@@ -809,6 +817,111 @@ public void fetchParticipationsWithSubmissionsAndResultsForCourses(List<Course>
         return ResponseEntity.ok().headers(HeaderUtil.createEntityDeletionAlert(applicationName, true, ENTITY_NAME, course.getTitle())).build();
     }
 
+    /**
+     * PUT /courses/{courseId} : archive an existing course asynchronously. This method starts the process of archiving all course exercises, submissions and results in a large
+     * zip file. It immediately returns and runs this task asynchronously. When the task is done, the course is marked as archived, which means the zip file can be downloaded.
+     *
+     * @param courseId the id of the course
+     * @return empty
+     */
+    @PutMapping("/courses/{courseId}/archive")
+    @PreAuthorize("hasAnyRole('ADMIN', 'INSTRUCTOR')")
+    public ResponseEntity<Void> archiveCourse(@PathVariable Long courseId) {
+        log.info("REST request to archive Course : {}", courseId);
+
+        // Get the course with all exercises
+        final Course course = courseService.findOneWithExercisesAndLectures(courseId);
+
+        final User user = userService.getUserWithGroupsAndAuthorities();
+        if (!authCheckService.isAtLeastInstructorInCourse(course, user)) {
+            throw new AccessForbiddenException("You are not allowed to access this resource");
+        }
+
+        // Archiving a course is only possible after the course is over
+        if (now().isBefore(course.getEndDate())) {
+            throw new BadRequestAlertException("You cannot archive a course that is not over.", ENTITY_NAME, "courseNotOver", true);
+        }
+
+        courseService.archiveCourse(course);
+
+        // TODOs Course Exercises
```

Comment:
Are this todos still relevant? In the description you mention that it supports programming, file upload, modeling and text exercises?